joomla/joomla-cms
# [4.1] poor man’s cron 3rd attempt

The last iteration of #25022, porting #22310 to 4.x

### Goal
a way to run scheduled "quick" tasks in approximately regular intervals

### Summary of Changes
- a new system plugin scheduler
- two new console command 

### How it works
In a normal hosting situation that would be job for cron, but if your hosting doesn’t offer cron, then a common solution in these cases is “poor man’s cron:” on every page load (`onAfterRespond`) the scheduler system plugin check if itself  is expected to run and  then trigger a new event `onExecuteScheduledTask`  then any job plugin wich listen to that event  check if itself  is expected to run and if so execute his task.  

### Testing Instructions (web)

- apply PR
- System -> database -> Update Structure  (there is a new table)
- System panel -> discover -> you'll find a new system plugin scheduler
- install it 
- set options, publish the plugin and save
for test only
Frequency (in minutes) set = 0 mean run every (`onAfterRespond`)

![Screenshot from 2020-06-14 08-13-58](https://user-images.githubusercontent.com/181681/84586143-22d3ad00-ae17-11ea-8ca0-4bf1d4be0451.png)


- install these 2 silly job plugin:
[plg_job_jobthree.zip](https://github.com/joomla/joomla-cms/files/4774301/plg_job_jobthree.zip)
[plg_job_jobone.zip](https://github.com/joomla/joomla-cms/files/4774283/plg_job_jobone.zip)

-  this one is an example on how to schedule an sql export
[plg_job_jobexportdb.zip](https://github.com/joomla/joomla-cms/files/4775936/plg_job_jobexportdb.zip)



-  job one simply sleep for 1 seconds		
-  job two simply sleep for 3 seconds

- set options, publish the plugin and save plugin ONE
Frequency (in unit) set 1
Unit of time set minute

 means that this plugin run 1xminute = every minute 
if `onExecuteScheduledTask`  has been triggered 

![Screenshot from 2020-06-14 08-16-11](https://user-images.githubusercontent.com/181681/84586168-68907580-ae17-11ea-9c4a-1a66bacd559b.png)


- set options, publish the plugin and save plugin TWO
maybe use different  settings than ONE

### Expected result
depending on your (plugins) settings 
take a coffe/smoke/whatever :smile: 
randomly click a couple of  frontend pages
check the log file at  `yoursite/administrator/logs/joomla_web.php`
you'll see something like this 
```php
#
#<?php die('Forbidden.'); ?>
#Date: 2019-05-30 17:24:39 UTC
#Software: Joomla! 4.0.0-alpha10-dev Development [ Amani ] 16-May-2019 19:22 GMT

#Fields: datetime	priority clientip	category	message
2019-05-30T17:24:39+00:00	INFO 127.0.0.1	scheduler	Scheduler>start
2019-05-30T17:24:40+00:00	INFO 127.0.0.1	scheduler	Job>jobone Task#221 Processing Time: 1.01 seconds.
2019-05-30T17:24:43+00:00	INFO 127.0.0.1	scheduler	Job>jobtwo Task#166 Processing Time: 3.01 seconds.
2019-05-30T17:24:43+00:00	INFO 127.0.0.1	scheduler	Scheduler>stop>task#1564 Total Processing Time: 4.04 seconds.
2019-05-30T17:24:47+00:00	INFO 127.0.0.1	scheduler	Scheduler>start
2019-05-30T17:24:47+00:00	INFO 127.0.0.1	scheduler	Scheduler>stop>task#1565 Total Processing Time: 0.04 seconds.
```

 **1st run** 
the scheduler plugin (wich runs always frequency=0) trigger the `onExecuteScheduledTask` event
so plugin ONE and plugin TWO  are executed (cause no previous run)

**2nd  run**
the scheduler plugin (wich runs always frequency=0) trigger the `onExecuteScheduledTask` event
but this time both plugin have an expected scheduled time (for example ONE each 1min TWO each 5min) 
the 2nd run was at 17:24:47+00:00	
the 1st run was at 17:24:43+00:00
they have only 4 seconds of difference form the previous execution so both plugin are not expected to run

**3rd  run** 
it depends on how fast you click & settings


### Testing Instructions (webcron)
on scheduler system plugin
set Webcron to yes/enable 
set Key to mywebcronactivationkey
then use `yoursiteurl/index.php?webcronkey=mywebcronactivationkey` as the url for webcron services to be triggred


### Testing Instructions (cli/cron)
Go to the cli folder of your website and run
`php joomla.php job:run` 


![Screenshot from 2020-06-13 10-35-22](https://user-images.githubusercontent.com/181681/84564406-adf26b80-ad61-11ea-979f-1666debc48fc.png)

check the log file at yoursite/administrator/logs/joomla_cli.php

`php joomla.php job:list`
![Screenshot from 2020-06-13 10-36-51](https://user-images.githubusercontent.com/181681/84564425-d8442900-ad61-11ea-9081-a21994512d07.png)

`php joomla.php job:run --help`
![Screenshot from 2020-06-13 10-42-48](https://user-images.githubusercontent.com/181681/84564541-bf884300-ad62-11ea-9701-d226130b660e.png)

### Actual result
N/A

### Additional comments

it can be used for tasks like:

- deleteExpiredConsents
- remindExpiringConsents
- logRotation
- updateNotifications
- sessionGC
- etc...

### Documentation Changes Required
yes
